include/mapbox/std/weak.hpp

```diff
@@ -20,14 +20,26 @@ class WeakPtrSharedData {
     WeakPtrSharedData() = default;
 
     void sharedLock() {
-        assert(valid());
-        sharedLocks_++;
+        std::size_t noLocks = sharedLocks_;
+        std::size_t incremented;
+        do {
+            if (noLocks == kInvalidValue) return;
+            incremented = noLocks + 1;
+            // `compare_exchange_weak()` is invoked in a cycle to handle the case,
+            // if another thread has just modified `sharedLocks_` (so that it is not
+            // equal to `noLocks` any more). We early return, if `sharedLocks_` became
+            // equal to `kInvalidValue`.
+        } while (!sharedLocks_.compare_exchange_weak(noLocks, incremented));
     }
 
     void sharedUnlock() {
-        assert(valid());
-        assert(sharedLocks_ > 0u);
-        sharedLocks_--;
+        std::size_t noLocks = sharedLocks_;
+        std::size_t decremented;
+        do {
+            if (noLocks == kInvalidValue) return;
+            decremented = noLocks - 1;
```

Comment:
This will overflow if `noLocks == 0`.

Reply:
(I realize that if used correctly, this situation will not occur, since a precondition of this function is that sharedLocks_ is > 0)

Reply:
should not happen, will fix the assertion below